src/java/org/apache/cassandra/io/compress/CompressionMetadata.java

```diff
@@ -181,79 +235,89 @@ public void addTo(Ref.IdentityCollection identities)
+
+        int chunkCount = endIndex - startIndex;
         Memory offsets = Memory.allocate(chunkCount * 8L);
-        int i = 0;
+        long i = 0;
         try
         {
-
+            input.skipBytes(startIndex * 8);
             for (i = 0; i < chunkCount; i++)
             {
-                offsets.setLong(i * 8L, input.readLong());
+                long value = input.readLong();
+                offsets.setLong(i * 8, value);
             }
-
             return offsets;
         }
+        catch (EOFException e)
+        {
+            offsets.close();
+            String msg = String.format("Corrupted Index File %s: read %d but expected %d chunks.", indexFilePath, i, chunkCount);
+            throw new CorruptSSTableException(new IOException(msg, e), indexFilePath);
+        }
         catch (IOException e)
         {
-            if (offsets != null)
-                offsets.close();
-
-            if (e instanceof EOFException)
-            {
-                String msg = String.format("Corrupted Index File %s: read %d but expected %d chunks.",
-                                           indexFilePath, i, chunkCount);
-                throw new CorruptSSTableException(new IOException(msg, e), indexFilePath);
-            }
+            offsets.close();
             throw new FSReadError(e, indexFilePath);
         }
     }
 
     /**
      * Get a chunk of compressed data (offset, length) corresponding to given position
      *
-     * @param position Position in the file.
+     * @param uncompressedDataPosition Position in the file.
      * @return pair of chunk offset and length.
      */
-    public Chunk chunkFor(long position)
+    public Chunk chunkFor(long uncompressedDataPosition)
     {
-        // position of the chunk
-        long idx = 8 * (position / parameters.chunkLength());
-
-        if (idx >= chunkOffsetsSize)
-            throw new CorruptSSTableException(new EOFException(), indexFilePath);
-
-        if (idx < 0)
-            throw new CorruptSSTableException(new IllegalArgumentException(String.format("Invalid negative chunk index %d with position %d", idx, position)),
-                                              indexFilePath);
+        int chunkIdx = chunkIndex(uncompressedDataPosition);
+        return chunk(chunkIdx);
+    }
 
-        long chunkOffset = chunkOffsets.getLong(idx);
-        long nextChunkOffset = (idx + 8 == chunkOffsetsSize)
-                                ? compressedFileLength
-                                : chunkOffsets.getLong(idx + 8);
+    private Chunk chunk(long chunkOffset, long nextChunkOffset)
+    {
+        return new Chunk(chunkOffset, Math.toIntExact(nextChunkOffset - chunkOffset - 4)); // "4" bytes reserved for checksum
+    }
 
-        return new Chunk(chunkOffset, (int) (nextChunkOffset - chunkOffset - 4)); // "4" bytes reserved for checksum
+    private Chunk chunk(int relChunkIdx)
```

Comment:
Could we expand on what "relative" means here?

Reply:
It is relative to the first chunk stored in the partial file / first chunk we read the offset for. Perhaps the "rel" is not needed in this context?